staging/src/kubevirt.io/api/clone/v1alpha1/types.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// VirtualMachineClone is a CRD that clones one VM into another.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +genclient
+type VirtualMachineClone struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   VirtualMachineCloneSpec   `json:"spec" valid:"required"`
+	Status VirtualMachineCloneStatus `json:"status,omitempty"`
+}
+
+type CloneIncludeExclude struct {
+	// +optional
+	CloneAllByDefault *bool `json:"cloneAllByDefault,omitempty"`
+	// +optional
+	// +listType=atomic
+	Exclude []string `json:"exclude,omitempty"`
+	// +optional
+	// +listType=atomic
+	Include []string `json:"include,omitempty"`
+}
+
+type VirtualMachineCloneSpec struct {
+	Source *corev1.TypedLocalObjectReference `json:"source"`
+
+	// +optional
+	Target *corev1.TypedLocalObjectReference `json:"target,omitempty"`
+
+	// +optional
+	Annotations *CloneIncludeExclude `json:"annotations,omitempty"`
```

Comment:
rather than this `CloneIncludeExclude` struct, what would you think about a list of strings that act as filters where we allow `*` to act as a wildcard.

so something like
```
AnnotationFilters []string `json:"annotationCloneFilters,omitempty"`
```

Copy everything with the `*` wildcard

```
annotationsFilters:
  - "*"
```

copy everything with the prefixes `myprefix.io` and `myotherprefix.io` using the `*` wildcard
```
annotationFilters:
- "myprefix.io/*"
- "myotherprefix.io/*"
```

copy explict keys without using wildcards.

```
annotationFilters:
- "myprefix.io/stuff"
```

And by default an empty list would just mean nothing is copied over.

Reply:
Very nice idea!
The only thing that's missing IMO is to express something like "copy everything except for keys with prefix ...". It sounds useful in order to get rid of certain metadata labels that aren't relevant for the target.

Maybe we can also use `!` to express that? Perhaps something like (that would mean to copy everything except for key with `irrelevantMetadata.io/` prefix):
```
annotationsFilters:
  - "*"
  - "!irrelevantMetadata.io/*"
```

But maybe this is becoming too complicated? This way we would also have to somehow verify that there are no conflicts, or at least determine a way to solve them.. WDYT?

Reply:
I like this in theory, but I wonder about the implementation.  Is a similar scheme use anywhere else in k8s?  I know wildcards are used in rbac rules but only to match everything.

Reply:
> Maybe we can also use ! to express that? Perhaps something like (that would mean to copy everything except for key with irrelevantMetadata.io/ prefix):

yeah that make makes sense.

> I like this in theory, but I wonder about the implementation. Is a similar scheme use anywhere else in k8s? I know wildcards are used in rbac rules but only to match everything. 

A simple implementation could use the golang strings pkg HasSuffix function (n^2) comparison operation, or compile a custom regex out of all the filters to do an (n) comparision operation.

I think we should make sure to limit the the functionality to the `!` character only being possible at the beginning of the filter to indicate it's a `negation` operation, and the `*` character only being allowed at the end of a filter to indicate it's a wildcard. So i wouldn't want someone trying this `*mykey/!something`. We just want people to be able to express include/exclude for the entire filter, and wildcard postfix matching.

Reply:
> I think we should make sure to limit the the functionality to the `!` character only being possible at the beginning of the filter to indicate it's a `negation` operation, and the `*` character only being allowed at the end of a filter to indicate it's a wildcard. So i wouldn't want someone trying this `*mykey/!something`. We just want people to be able to express include/exclude for the entire filter, and wildcard postfix matching.

Makes sense. Invalid use of `!` and `*` can be rejected at validating webhooks.

BTW, I think that usage such as `kubevirt.io/*/something` might also be useful in the future, but I agree that as a first step we should allow the most simple use only.